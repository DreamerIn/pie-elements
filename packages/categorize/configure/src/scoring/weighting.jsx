import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Panel from './panel';
import { NumberTextField } from '@pie-lib/config-ui';
import grey from '@material-ui/core/colors/grey';
import Chip from '@material-ui/core/Chip';
import debug from 'debug';
const log = debug('@pie-element:categorize:configure:weighting');
import Grid, { Cell } from './grid';

const Percent = ({ percent, className }) => (
  <Chip classes={{ root: className }} label={`${percent}%`} />
);

Percent.propTypes = {
  percent: PropTypes.string.isRequired,
  className: PropTypes.string
};

const cell = () => ({
  alignSelf: 'center',
  borderBottom: `solid 1px ${grey[300]}`
});

const Rule = withStyles(() => ({
  label: cell(),
  number: cell(),
  percent: cell()
}))(({ classes, rule, onChange, percent, disabled }) => {
  const changeRule = (event, number) => {
    rule.points = number;
    onChange(rule);
  };
  return (
    <React.Fragment>
      <Cell html={rule.label} />
      <Cell>
        <NumberTextField
          disabled={disabled}
          className={classes.number}
          min={0}
          max={20}
          value={rule.points}
          suffix={rule.points === 1 ? 'pt' : 'pts'}
          onChange={changeRule}
        />
      </Cell>
      <Cell>
        <Percent className={classes.percent} percent={percent} />
      </Cell>
    </React.Fragment>
  );
});

const build = (rules, categories) => {
  return categories.map(c => {
    const rule = rules.find(r => r.category === c.id) || { points: 1 };
    return {
      label: c.label,
      category: c.id,
      points: rule.points
    };
  });
};

export class Weighting extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    weighting: PropTypes.object,
    onChange: PropTypes.func.isRequired,
    categories: PropTypes.array
  };
  static defaultProps = {};

  toggleEnabled = () => {
    const { weighting, onChange } = this.props;
    weighting.enabled = !weighting.enabled;
    onChange(weighting);
  };

  changeRule = rule => {
    const { weighting, onChange } = this.props;

    const index = weighting.rules.findIndex(r => r.category === rule.category);
    const r = { category: rule.category, points: rule.points };
    if (index === -1) {
      weighting.rules.push(r);
    } else {
      weighting.rules.splice(index, 1, r);
    }

    log('[changeRule]: ', weighting);

    onChange(weighting);
  };

  getPercent = (rule, rules) => {
    const total = rules.reduce((sum, r) => sum + r.points, 0);
    const rawPercent = rule.points / total * 100;
    return rawPercent.toFixed(0);
  };

  render() {
    const { weighting, categories } = this.props;

    const rules = build(weighting.rules, categories);
    return (
      <Panel
        title="Weighting"
        enabled={weighting.enabled}
        onToggleEnabled={this.toggleEnabled}
      >
        <Grid headings={['Category', 'Points', '']} size={'1fr 70px 70px'}>
          {rules.map((r, index) => (
            <Rule
              rule={r}
              disabled={!weighting.enabled}
              percent={this.getPercent(r, rules)}
              onChange={this.changeRule}
              key={index}
            />
          ))}
        </Grid>
      </Panel>
    );
  }
}
const styles = () => ({
  weighting: {},
  ruleHolder: {
    display: 'grid',
    backgroundColor: 'red',
    gridTemplateColumns: '1fr 70px 70px',
    gridRowGap: '1px',
    gridColumnGap: 0
  }
});
export default withStyles(styles)(Weighting);
