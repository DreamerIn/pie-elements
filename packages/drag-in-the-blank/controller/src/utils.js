import reduce from 'lodash/reduce';
import find from 'lodash/find';

export const getAllCorrectResponses = ({ correctResponse, alternateResponses }) => {
  return reduce(correctResponse || {}, (obj, val, key) => {
    obj[key] = [val];

    if (alternateResponses && alternateResponses[key]) {
      obj[key] = [
        ...obj[key],
        ...alternateResponses[key]
      ];
    }

    return obj;
  }, {});
};
