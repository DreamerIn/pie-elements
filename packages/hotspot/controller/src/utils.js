import isEqual from 'lodash/isEqual';
import isEmpty from 'lodash/isEmpty';

export const getCorrectResponse = (choices) => choices
  .filter(c => c.correct)
  .map(c => ({ id: c.id }))
  .sort();

export const isResponseCorrect = (question, session) => {
  const { shapes: { rectangles, polygons } } = question;
  const choices = [...rectangles, ...polygons];
  let correctResponse = getCorrectResponse(choices);

  if (!session || isEmpty(session)) {
    return false;
  }

  if (session.answers && session.answers.length) {
    return isEqual((session.answers || []).sort(), correctResponse);
  } else if (!(correctResponse && correctResponse.length)) {
    return true;
  }
  return false;
};
