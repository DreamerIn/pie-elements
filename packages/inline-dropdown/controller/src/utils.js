import isEqual from 'lodash/isEqual';
import reduce from 'lodash/reduce';
import find from 'lodash/find';

export const isResponseCorrect = (question, session) => {
  if (!question || !session || !session.value) {
    return false;
  }

  let correctResponse = reduce(question.choices, (acc, area, key) => {
    const correctChoice = find(area, c => c.correct);

    acc[key] = correctChoice.value;

    return acc;
  }, {});

  return isEqual((session.value || {}), correctResponse);
};
