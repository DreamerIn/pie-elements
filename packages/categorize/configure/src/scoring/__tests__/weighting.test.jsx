import { shallow } from 'enzyme';
import React from 'react';

import { Weighting } from '../weighting';
import { callsOnChange as _callsOnChange } from '../../__tests__/utils';

describe('Weighting', () => {
  let w;
  let onChange = jest.fn();
  const wrapper = extras => {
    const defaults = {
      classes: {},
      className: 'className',
      onChange,
      categories: [{ id: '1', label: 'Content' }],
      weighting: {
        enabled: true,
        rules: [{ category: '1', points: 1 }]
      }
    };
    const props = { ...defaults, ...extras };
    return shallow(<Weighting {...props} />);
  };

  describe('snapshot', () => {
    it('renders', () => {
      w = wrapper();
      expect(w).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    const callsOnChange = function() {
      const args = Array.prototype.slice.call(arguments);
      const first = args.shift();
      const startArgs =
        typeof first === 'string' ? [wrapper(), first] : [first];

      _callsOnChange.apply(null, startArgs.concat(args));
    };

    describe('toggleEnabled', () => {
      callsOnChange('toggleEnabled', () => {
        expect(onChange).toBeCalledWith(
          expect.objectContaining({ enabled: false })
        );
      });
      callsOnChange(
        wrapper({ weighting: { enabled: false, rules: [] } }),
        'toggleEnabled',
        () => {
          expect(onChange).toBeCalledWith(
            expect.objectContaining({ enabled: true })
          );
        }
      );
    });

    describe('changeRule', () => {
      callsOnChange('changeRule', { category: '1', points: 2 }, () => {
        expect(onChange).toBeCalledWith(
          expect.objectContaining({
            rules: [
              {
                category: '1',
                points: 2
              }
            ]
          })
        );
      });
    });

    describe('getPercent', () => {
      it('returns the percent', () => {
        const result = wrapper()
          .instance()
          .getPercent({ points: 1 }, [{ points: 1 }, { points: 1 }]);
        expect(result).toEqual('50');
      });
    });
  });
});
