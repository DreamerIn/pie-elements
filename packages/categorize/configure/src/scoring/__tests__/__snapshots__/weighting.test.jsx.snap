// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weighting snapshot renders 1`] = `
<WithStyles(Panel)
  enabled={true}
  onToggleEnabled={[Function]}
  title="Weighting"
>
  <WithStyles(Grid)
    headings={
      Array [
        "Category",
        "Points",
        "",
      ]
    }
    size="1fr 70px 70px"
  >
    <WithStyles(Component)
      disabled={false}
      key="0"
      onChange={[Function]}
      percent="100"
      rule={
        Object {
          "category": "1",
          "label": "Content",
          "points": 1,
        }
      }
    />
  </WithStyles(Grid)>
</WithStyles(Panel)>
`;
